### PR TITLE
Wsp/locale source

### PR DESCRIPTION
**Pull request by wpride**
@phillipm my take on your fix

### CODE REVIEW

File: app/src/org/commcare/utils/AndroidArrayDataSource.java [hunk 1]
```diff
@@ -18,12 +20,11 @@ public AndroidArrayDataSource(Context context){
 
     @Override
     public String[] getArray(String key) {
-        int resourceId = context.getResources().getIdentifier(key, "array", context.getPackageName());
-
-        if (resourceId == 0) {
-            throw new RuntimeException("Localized array data for '" + key + "' not found");
+        if(key.contains("ethiopian")){
```
Comment (phillipm) [inline app/src/org/commcare/utils/AndroidArrayDataSource.java]: The imprecision of this worries me. For instance, what if we have 2 ethiopian related arrays: days and months?

File: app/src/org/commcare/views/widgets/WidgetFactory.java [hunk 2]
```diff
@@ -92,6 +94,8 @@ private QuestionWidget buildBasicWidget(String appearance, FormEntryPrompt fep,
             case Constants.DATATYPE_DATE_TIME:
                 return new DateTimeWidget(context, fep);
             case Constants.DATATYPE_DATE:
+                // Need to override CalendarUtil's month localizer
+                CalendarUtils.setArrayDataSource(new AndroidArrayDataSource(context));
```
Comment (phillipm) [inline app/src/org/commcare/views/widgets/WidgetFactory.java]: 👍

File: unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import org.robolectric.Shadows;
+import org.robolectric.annotation.Config;
+import org.robolectric.shadows.ShadowActivity;
+import org.robolectric.shadows.ShadowEnvironment;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertTrue;
+
+/**
+ * @author Phillip Mates (pmates@dimagi.com).
```
Comment (phillipm) [inline unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java]: 😄

File: unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                        .create().start().resume().get();
+
+        ImageButton nextButton = (ImageButton)formEntryActivity.findViewById(R.id.nav_btn_next);
+
+        // enter an answer for the question
+        TextView dayText = (TextView)formEntryActivity.findViewById(R.id.daytxt);
+        TextView monthText = (TextView)formEntryActivity.findViewById(R.id.monthtxt);
+        TextView yearText = (TextView)formEntryActivity.findViewById(R.id.yeartxt);
+
+        assertEquals(monthText.getText(), "Ashadh");
```
Comment (phillipm) [inline unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java]: Do you think there is a way to change the locale in robolectric tests and check that the month names changed from `Ashadh` to `आषाढ`?

Comment (phillipm) [inline unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java]: Looks like there is a way to change locales in robolectric tests. Can you include this in your test?

**Comment by phillipm** [comment]
Heads up that this is targeting `master` not `commcare_2.29`

File: unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java (hunk 4 above)
Comment (wpride) [inline unit-tests/src/org/commcare/android/tests/formnav/CalendarLocaleTest.java]: Wasn't able to get this working - tried the @Config annotation as well as the configuration.setLocale() call.

**Comment by amstone326** [comment]
@phillipm I haven't reviewed this because it seems like you were in the process of doing so. Would you be able to take another look before code freeze?

**Comment by wpride** [comment]
@phillipm @amstone326 reverted my change to AndroidArrayDataSource, think I prefer phil's So this is just the test and moving the CalendarUtils instantiation call now.

**Pull request merged**